### ISSUE TITLE
Define documentation structure

### ISSUE THREAD

**Issue by fd0** [restic/restic#1273]
Referring to a comment @jniggemann made in https://github.com/restic/restic/issues/664#issuecomment-330615622 I'd like to define the structure and contents of the documentation we have for restic.

**Edit: This issue is more about the structure and plans for the documentation, not so much about the implementation.**

At the moment there are several places where documentation is kept:
 * `README.md`
 * The manual
 * The help texts/manpages (the manpages are generated from the source code)
 * `CONTRIBUTING.md`
 
My idea is to restructure/reword the documents as follows:
 * `README.md`: Contains just the bare minimum: One to two paragraphs what the project is about, how to build it and make a backup, then links to other important places (manual, changelog, license)
 * The help texts/manpages: They should explain every command in detail. Some of the text (e.g. for the `forget` command) is contained in the manual ([here](https://github.com/restic/restic/blob/master/doc/manual.rst#removing-old-snapshots)). I'm not sure what to do with the longer explanatory text for removing snapshots, leave it in the manual? And include a shortened version in the manpage?
 * The manual: Should be refactored as a "book" style text that you can read from top to bottom, not so much as a reference. Maybe we can publish the manpages as markdown on readthedocs.io (Cobra can generate markdown from the help texts) and link to the appropiate places.
 * `CONTRIBUTING.md`: Should explain the development process and submitting patches (what's expected, format of the commits etc.). The document is way too long at the moment, this needs to be shortened.
 
Any other ideas on what to improve?

**Comment by rumpelsepp** [comment]
> The help texts/manpages: They should explain every command in detail. Some of the text (e.g. for the forget command) is contained in the manual (here). I'm not sure what to do with the longer explanatory text for removing snapshots, leave it in the manual? And include a shortened version in the manpage?

I think it is even possible to cover the restic documentation completely with manpages. The advantage is, that I have them always around even without internet connection. The Git project is documented only via manpages; for general topics you can create a, e.g., `restic-tutorial(7)` manpage, see [gittutorial(7)](https://www.kernel.org/pub/software/scm/git/docs/gittutorial.html).

I can recommend asciidoctor for this, as it renders manpages nicely. Setting up a travis job which pushes to github-pages is trivial. It makes sense to create a `restic` manpage which can also be shown as the README on github which links to all other available pages. This way you have everything available online and offline in a (kind of) standardized format.

**Comment by wscott** [comment]
BTW I recently accidentally discovered the manpages. The 'install' section of the manual doesn't actually say how to install restic at all, just how to build it. And so it doesn't mention that installing the manpages would be a good idea.

Ideally, the similar build from source instructions should be in your README file (or a separate INSTALL file) and not in the book-style documentation which is harder to find from the command line.

**Comment by jniggemann** [comment]
### `README.md`
I think it's a good idea to have it as short as possible and concur with the contents fd0 describes. Only difference I'd make: Having the license information in the `README.md` is a must and also encouraged by github (although I forgot where they say that).

### help texts/manpages
I don't have ideas for them yet, as I'd first like to work on the manual.

### manual
>Should be refactored as a "book" style text that you can read from top to bottom

I fully agree with that! I wanted to work on it a couple hours more, but as we're already on the subject:  
Yesterday, I drafted an outline for a new manual. Look [here ](https://github.com/jniggemann/restic/wiki)and please give me feedback. I think that we could re-use a lot of the existing documentation and add new stuff where needed. I'd like to do this and would like to proceed as follows:
- Gather feedback on document structure from other restic users
- Re-write the manual chapter by chapter
- publish to my own webspace ASAP to gather feedback on chapter
- next chapter, same procedure

**Sidenote**: I'm the German translator of the `obnam` manual ([English](http://code.liw.fi/obnam/manual/obnam-manual.en.pdf), [German](http://code.liw.fi/obnam/manual/obnam-manual.de.pdf)) and found the structure of the document really good, so it doesn't come as a surprise that my draft closely resembles that structure.

### CONTRIBUTING.md
It's a bit long indeed, but other than that, not yet any ideas from my side.

**Comment by rumpelsepp** [comment]
I was referring to restic being installed via a package management system, such as `apt`. 

>  and not in the book-style documentation which is harder to find from the command line

Maybe some link in `restic --help` which refers to the online version of the documentation?

**Comment by jniggemann** [comment]
@rumpelsepp: I like asciidoctor very much, but sphinx can already build manpages ([output formats](http://www.sphinx-doc.org/en/stable/man/sphinx-build.html)) so we should IMHO stick with what we already have.

That said, what do you think about re-using single chapters of the manual as the manpages?

**Comment by fd0** [comment]
Ah, I should've pointed that out in the first comment: This issue is more about the structure and plans for the documentation, not so much about the implementation.

For the manual and the other texts in `doc/` (e.g. the design document) we'll stick with Sphinx (and readthedocs) for now.

For the manpages I like that Cobra can generate them so they're always up to date, reflect the options as they are implemented and it's the same text that is displayed when `--help` is called. I'd like to expand the texts in the command definitions that will show up in the help and the man pages.

For example, this is the definition of the `backup` command:
https://github.com/restic/restic/blob/24c62e719a0781956e723dee7d857c496a4cbfd3/cmd/restic/cmd_backup.go#L22-L27

The text will show up in the manual page:
https://github.com/restic/restic/blob/24c62e719a0781956e723dee7d857c496a4cbfd3/doc/man/restic-backup.1#L18-L19

And in the help text:
```
$ restic backup --help
The "backup" command creates a new snapshot and saves the files and directories
given as the arguments.

Usage:
  restic backup [flags] FILE/DIR [FILE/DIR] ...
[...]
```

**Comment by rumpelsepp** [comment]
> That said, what do you think about re-using single chapters of the manual as the manpages?

Of course that also works. But in my experience the sphinx manpage backend is a bit buggy and often produces strange looking manpages (when viewing them in the terminal). But yeah, the result according to the available docs should be the same.

**Comment by rumpelsepp** [comment]
> For the manpages I like that Cobra can generate them so they're always up to date, reflect the options as they are implemented and it's the same text that is displayed when --help is called. I'd like to expand the texts in the command definitions that will show up in the help and the man pages.

Is it possible with cobra to generate manpages that include more information, e.g. an additional section? Or is this limited to just a short description of available options? Too sparse manpages are often useless, since in question you are forced to open the browser and read the full description in the book or whatever anyway.

A good manpage should at least include some useful examples and the respected environment variables (if present). I often grep the restic documentation in order to find out if it was `RESTIC_PASSWORD` or something else. I would expect to find this kind of information in `restic-backup(1)` in the "Environment Variables" section.

**Comment by fd0** [comment]
> Is it possible with cobra to generate manpages that include more information, e.g. an additional section? Or is this limited to just a short description of available options?

The text can be of arbitrary length, and will make up most of the manpage. I'll need to check the code to see if there's anything else. Maybe even sub-sections are possible...

> I would expect to find this kind of information in restic-backup(1) in the "Environment Variables" section.

Good idea.

**Comment by fd0** [comment]
I've just tested it, we can use markdown syntax and it will correctly render in the manpage, e.g. as a headline/section.

**Comment by jniggemann** [comment]
I'm working on this

**Comment by rawtaz** [comment]
@fd0 Can this issue be closed, or did the PR (#1300) not resolve it entirely?

**Comment by rawtaz** [comment]
I'm closing this, from what I can tell most of this is pretty much met by the current documentation structure we have. That one is still up for improvement, both short- and long-term, which is an ongoing task. If someone things there is something that needs fixing with the current documentation, please open a new issue for that. Thank you! 🚀
